src/pathfinder.rs

```diff
@@ -0,0 +1,266 @@
+        return Err(DeleteNodeError::NodeNotFound);
+    }
+    let node_to_delete = node_to_delete.unwrap().clone().unwrap();
+    node_to_delete.connected_nodes_id.iter().for_each(|i| {
+        if let Some(Some(node)) = nodes.get_mut(*i) {
+            node.connected_nodes_id.retain(|id| id != i)
+        }
+    });
+
+    let node = nodes.get_mut(unique_id).unwrap();
+    *node = None;
+    Ok("1".to_string())
+}
+
+pub enum AstarError {
+    MutexError(mut_static::Error),
+    StartNodeNotFound,
+    GoalNodeNotFound,
+    NoPathFound,
+}
+
+impl From<mut_static::Error> for AstarError {
+    fn from(error: mut_static::Error) -> Self {
+        AstarError::MutexError(error)
+    }
+}
+
+byond_fn!(fn astar_generate_path(start_node_id, goal_node_id) {
+    if let (Ok(start_node_id), Ok(goal_node_id)) = (start_node_id.parse::<usize>(), goal_node_id.parse::<usize>()) {
+        match astar_generate_path_(start_node_id, goal_node_id) {
+            Ok(vector) => Some(match serde_json::to_string(&vector) {
+                Ok(s) => s,
+                Err(_) => "Cannot serialize path".to_string(),
+            }),
+            Err(e) => Some(match e {
+                AstarError::MutexError(e) => format!("Mutex error : {}", e),
+                AstarError::StartNodeNotFound => "Start node not found".to_string(),
+                AstarError::GoalNodeNotFound => "Goal node not found".to_string(),
+                AstarError::NoPathFound => "No path found".to_string(),
+            })
+        }
+    }
+    else {
+        Some("Invalid arguments".to_string())
+    }
+});
+
+fn astar_generate_path_(
+    start_node_id: usize,
+    goal_node_id: usize,
+) -> Result<Vec<usize>, AstarError> {
+    let nodes = NODES.read()?;
+
+    let start_node = nodes.get(start_node_id);
+    if start_node.is_none() || start_node.unwrap().is_none() {
+        return Err(AstarError::StartNodeNotFound);
+    }
+
+    let goal_node = nodes.get(goal_node_id);
+    if goal_node.is_none() || goal_node.unwrap().is_none() {
+        return Err(AstarError::GoalNodeNotFound);
+    }
+
+    if start_node.unwrap().as_ref().unwrap().z != goal_node.unwrap().as_ref().unwrap().z {
+        return Err(AstarError::NoPathFound);
+    }
+
+    let path = astar(
+        &start_node.unwrap().clone().unwrap(),
+        |node| node.successors(),
+        |node| node.distance(&goal_node.unwrap().clone().unwrap()),
+        |node| node == &goal_node.unwrap().clone().unwrap(),
+    );
+    if path.is_none() {
+        return Err(AstarError::NoPathFound);
+    }
+
+    let (path, _) = path.unwrap();
+    Ok(path.into_iter().map(|node| node.unique_id).collect())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[allow(unused_must_use)]
+    fn test_register_node() {
+        let json = std::fs::read_to_string("tests/rsc/ai_nodes_info.json").unwrap();
+        register_nodes_(&json).is_ok();
+        assert!(NODES.read().unwrap().len() != 0)
```

Comment:
when asserting register nodes directly, it fails 50% of the time. I don't know why, if anyone has an idea please shoot